### PR TITLE
Acknowledge each chunk and async upload validation

### PR DESCRIPTION
**Pull request by luffs**
Acknowledge each chunks to make the progress event accurate. Also uses
less memory when uploading big files, since it avoids reading the entire
file into memory.

Added uploadValidator to the server Uploader object, that can be overridden to enable async
preparing and validation.

### CODE REVIEW

File: server.js [hunk 1]
```diff
@@ -256,6 +263,11 @@ function SocketIOFileUploadServer() {
 								memo: "self-thrown from progress event"
 							});
 						}
+                        else {
```
Comment (luffs) [inline server.js]: Ah, I only use the the auto-saving feature, so didn't think about that. Fixed in new commit.

**Comment by luffs** [comment]
Ah, I only use the the auto-saving feature, so didn't think about that. Fixed in new commit, also updated the readme and fixed the indents.

**Comment by vote539** [comment]
Thanks for the update!  I've been super busy the last couple of days so this weekend I'll find time to review and merge the PR.  :simple_smile:

**Comment by luffs** [comment]
No worries, thanks for making the library! Made another commit that fixes an bug when there are multiple instances of the client library in the browser.

**Comment by Sjiep** [comment]
This PR looks great, looking forward for it to be merged!

**Comment by vote539** [comment]
I've merged your patch!  Very sorry for the delay.  I made two changes to your code:

1. Fixed typo on line 584, `readyCallbacks` vs. `chunkCallbacks`
2. You had the uploadValidator running only when `uploader.dir` was set.  I changed it such that uploadValidator always runs regardless of the value of `uploader.dir`.

I also noticed in your README example that you're changing `uploader.dir` on the fly in uploadValidator.  This can result in a race condition.  I created #62 to address this issue.

Thanks again for the contribution, and have a happy new year!

**Pull request merged**